### PR TITLE
Workflow: Completion Service sample data / metadata

### LINKED ISSUE #255: Workflow: CompletionService attach per sample metadata/results
relies on #253 

## Summary
Allow SamplesWorkflowsExecution to have attachments

Add the following functionality to the Workflow Completion Service

Add per sample metadata/output to SamplesWorkflowsExecution models on the WorkflowExecution 

## References
https://github.com/phac-nml/pipeline-standards?tab=readme-ov-file#422-complete-example
"output/irida.output.json.gz"

we are not allowing nested/complex metadata, like this
https://github.com/phac-nml/pipeline-standards?tab=readme-ov-file#4241-simplified-metadata-json

### PR DESCRIPTION
**Pull request by JeffreyThiessen**
## What does this PR do and why?
_Describe in detail what your merge request does and why._
Fixes #255

* Modifies WorkflowExecutions:CompletionService to include
  * sample output files on associated SamplesWorkflowExecutions
  * sample output metadata on associated SamplesWorkflowExecutions
  * major refactor to break logic into smaller functions
* Updates `samples_workflow_executions` table to include `metadata` column as `jsonb`
* Updates SamplesWorkflowExecution model to have `outputs` as `:attachable`

Note: saving attachments/metadata will be in this issue https://github.com/phac-nml/irida-next/issues/495
Note: setting up completion jobs and reworking workflow execution states will be in this issue https://github.com/phac-nml/irida-next/issues/437
Note: This PR assumes the iridanext.output.json.gz file will always be formed correctly and the samples/samples_workflow_executions referenced exist.
iridanext.output.json.gz integrety validation will be handled in this issue: https://github.com/phac-nml/irida-next/issues/497

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._

#### Note: When testing with ga4gh connection + local DB for blob storage, the ga4gh_wes connection will not write files back to your db, so you must manually place the files where they are expected to be. If you are using Azure for blob storage with ga4gh_wes hosted on Azure, this will be handled for you like in prod.

1. Run a workflow execution via GUI
- you should upload a paired end to a new sample and use that sample as your workflow input
image

2. wait for state to become `completed`
image

3. Get your WorkflowExecution in rails console
```bash
rails c
```
```ruby
# Workflow Execution
wfe = WorkflowExecution.last
```

4. Local DB step: get your output directory
```ruby
# This is where the workflow execution data should be
wfe.workflow_params['--outdir']

# example:
# "/home/jthiessen/git/irida-next-core/storage/6e/zc/6ezc87o51ym3y4xtljonlucelzuz/output/"
```
You will need to create the `/output/` sub directory to place files in

Use the files in `test/fixtures/files/blob_outputs/normal2` or make your own
first `cd` command should match above without the `output/`
```bash
cd ~/git/irida-next-core/storage/aa/bb/genkey/
mkdir output
cd output
# note: iridanext.output.json.gz must be gzipped, files in test/fixtures are not gzipped
cp ~/git/irida-next-core/test/fixtures/files/blob_outputs/normal2/iridanext.output.json.gz .
cp ~/git/irida-next-core/test/fixtures/files/blob_outputs/normal2/summary.txt .
cp ~/git/irida-next-core/test/fixtures/files/blob_outputs/normal2/analysis1.txt .
cp ~/git/irida-next-core/test/fixtures/files/blob_outputs/normal2/analysis2.txt .
cp ~/git/irida-next-core/test/fixtures/files/blob_outputs/normal2/analysis3.txt .
```

5. Run the Completion Service
```ruby
WorkflowExecutions::CompletionService.new(wfe).execute
```

6. Check that the files are attached to the SamplesWorkflowExecutions on our WorkflowExecution
```ruby
wfe.workflow_samples_executions.count
>2

wfe.workflow_samples_executions[0].outputs.count
>2

wfe.workflow_samples_executions[0].outputs[0].filename
>#<ActiveStorage::Filename:0x00007f9936da89b0 @filename="analysis1.txt">

wfe.samples_workflow_executions[0].outputs[1].filename
#<ActiveStorage::Filename:0x00007f9936deff90 @filename="analysis2.txt">

wfe.workflow_samples_executions[1].outputs.count
>1

wfe.samples_workflow_executions[1].outputs[0].filename
#<ActiveStorage::Filename:0x00007f9936df5490 @filename="analysis3.txt">
```

7. Check that the metadata has been added to the SamplesWorkflowExecutions on our WorkflowExecution
```ruby
wfe.samples_workflow_executions[0].metadata
>{"number"=>1, "organism"=>"an organism"}

wfe.samples_workflow_executions[1].metadata
>{"number"=>2, "organism"=>"a different organism"}
```

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.

- [x] I have evaluated the [PR acceptance checklist](https://phac-nml.github.io/irida-next/docs/development/development_processes/code_review#acceptance-checklist) for this PR.

### CODE REVIEW

File: test/services/workflow_executions/completion_service_test.rb [hunk 1]
```diff
@@ -4,67 +4,178 @@
...
+      @normal2_output_analysis2_file_blob = make_and_upload_blob(
+        filepath: 'test/fixtures/files/blob_outputs/normal2/analysis2.txt',
+        blob_run_directory: blob_run_directory_c
+      )
+      @normal2_output_analysis3_file_blob = make_and_upload_blob(
+        filepath: 'test/fixtures/files/blob_outputs/normal2/analysis3.txt',
+        blob_run_directory: blob_run_directory_c
+      )
+
+      # override puid's so we can use the ones in the prepared iridanext.output.json.gz
```
[the remark this answers is not in the document]
Comment (JeffreyThiessen) [reply test/services/workflow_executions/completion_service_test.rb]: Fixed in https://github.com/phac-nml/irida-next/pull/487/commits/8aa1b8a6eb78392416fadd99d1a3ef29147e6df2

File: app/services/workflow_executions/completion_service.rb [hunk 2]
```diff
@@ -38,32 +42,79 @@ def execute
...
+
+        @attachable_blobs_tuple_list.append({ attachable: samples_workflow_execution,
+                                              blob_id_list: sample_file_blob_list })
+      end
+    end
+
+    def get_samples_workflow_executions_by_sample_puid(puid:)
+      @workflow_execution.samples_workflow_executions \
+                         .joins(:sample) \
+                         .where(sample: { puid: }) \
```
[the remark this answers is not in the document]
Comment (JeffreyThiessen) [reply app/services/workflow_executions/completion_service.rb]: Fixed in https://github.com/phac-nml/irida-next/pull/487/commits/ae5f936fbbe53d6eb857a3ca6370440fd0e814a2

**Review by ericenns — APPROVED** [review]
Looks good to me!

**Review by deepsidhu85 — APPROVED** [review]
Looks good to me!

**Pull request merged**